### ISSUE TITLE
Multi Hub Support?

### ISSUE THREAD

**Issue by m4d-maNu** [materik/homebridge-harmonyhub-plugin#2]
Hi,
I have 2 Hubs, but the plugin find only 1.

Is the Multi Hub Support included from the old Plugin?

**Comment by materik** [comment]
@m4d-maNu could you provide the log for when you start up your homebridge with the -D flag?

**Comment by m4d-maNu** [comment]
`sudo homebridge -D -U ~/.homebridge-harmony
[2018-10-30 13:36:00] Loaded config.json with 0 accessories and 1 platforms.
[2018-10-30 13:36:00] ---
[2018-10-30 13:36:01] Loaded plugin: homebridge-harmonyhub
[2018-10-30 13:36:01] Registering platform 'homebridge-harmonyhub.HarmonyHub'
[2018-10-30 13:36:01] ---
[2018-10-30 13:36:01] Loading 1 platforms...
[2018-10-30 13:36:01] [HarmonyHub] Initializing HarmonyHub platform...
[2018-10-30 13:36:01] [HarmonyHub] {Plugin}
[2018-10-30 13:36:01] Loading 0 accessories...
[2018-10-30 13:36:01] [HarmonyHub] {Accessory} fromAccessory
[2018-10-30 13:36:01] [HarmonyHub] {Accessory} <578eae74-6368-4f4c-861f-0089838a4eb7> <Schlafzimmer-HarmonyHub>
[2018-10-30 13:36:01] [HarmonyHub] {HarmonyHubAccessory} fromAccessory
[2018-10-30 13:36:01] [HarmonyHub] {HarmonyHubAccessory} <578eae74-6368-4f4c-861f-0089838a4eb7> <Schlafzimmer-HarmonyHub>
[2018-10-30 13:36:01] [HarmonyHub] {Plugin} configureAccessory: <Schlafzimmer-HarmonyHub>
Setup Payload:
X-HM://##########
Scan this code with your HomeKit app on your iOS device to pair with Homebridge:
                                
                                                      
Or enter this code with your HomeKit app on your iOS device to pair with Homebridge:
                       
    ┌────────────┐     
    │ xxx-xx-xxx │     
    └────────────┘     
                       
[2018-10-30 13:36:01] [HarmonyHub] {Plugin} didFinishLaunching
[2018-10-30 13:36:01] [HarmonyHub] {Plugin} loadAccessories
[2018-10-30 13:36:01] [HarmonyHub] {HarmonyHub} finding hubs
[2018-10-30 13:36:01] Homebridge is running on port 51880.
[2018-10-30 13:36:02] [HarmonyHub] {HarmonyHub} Found <1> hubs
[2018-10-30 13:36:02] [HarmonyHub] {HarmonyHub} <578eae74-6368-4f4c-861f-0089838a4eb7> <Schlafzimmer-HarmonyHub>
[2018-10-30 13:36:02] [HarmonyHub] {Plugin} use cache for <Schlafzimmer-HarmonyHub>
[2018-10-30 13:36:02] [HarmonyHub] {Service} <34615073> <AirPlay Audio>
[2018-10-30 13:36:02] [HarmonyHub] {Service} <23348175> <Television>
[2018-10-30 13:36:02] [HarmonyHub] {Service} <36078732> <Amazon Fire TV>
[2018-10-30 13:36:02] [HarmonyHub] {Service} <32221096> <Plex>
[2018-10-30 13:36:02] [HarmonyHub] {Service} <32686239> <AirPlay Mirroring>
[2018-10-30 13:36:02] [HarmonyHub] {HarmonyHubAccessory} found <5> services
[2018-10-30 13:36:02] [HarmonyHub] {HarmonyHub} create new client for <Schlafzimmer-HarmonyHub>
[2018-10-30 13:36:02] [HarmonyHub] {HarmonyHub} get activities
[2018-10-30 13:36:03] [HarmonyHub] {Activity} <34615073> <AirPlay Audio>
[2018-10-30 13:36:03] [HarmonyHub] {Activity} <23348175> <Television>
[2018-10-30 13:36:03] [HarmonyHub] {Activity} <36078732> <Amazon Fire TV>
[2018-10-30 13:36:03] [HarmonyHub] {Activity} <32221096> <Plex>
[2018-10-30 13:36:03] [HarmonyHub] {Activity} <32686239> <AirPlay Mirroring>
[2018-10-30 13:36:03] [HarmonyHub] {HarmonyHub} Found <5> activities: <AirPlay Audio, Television, Amazon Fire TV, Plex, AirPlay Mirroring>
[2018-10-30 13:36:03] [HarmonyHub] {HarmonyHubAccessory} update states for <5> services
[2018-10-30 13:36:03] [HarmonyHub] {HarmonyHubAccessory} handle state changes for <5> services
[2018-10-30 13:36:03] [HarmonyHub] {Service} set <AirPlay Audio> <off>
[2018-10-30 13:36:03] [HarmonyHub] {Service} set <Television> <off>
[2018-10-30 13:36:03] [HarmonyHub] {Service} set <Amazon Fire TV> <off>
[2018-10-30 13:36:03] [HarmonyHub] {Service} set <Plex> <off>
[2018-10-30 13:36:03] [HarmonyHub] {Service} set <AirPlay Mirroring> <off>`

**Comment by materik** [comment]
Hmm. I rely on another npm package to fetch the hubs so if it doesn't find any more I cannot do much about it. Try removing the cached accessories maybe.

**Comment by m4d-maNu** [comment]
> Try removing the cached accessories maybe.

I trying, but only find 1 Hub.

**Comment by UrsLebowski** [comment]
Hello, I can report the same. I've distributed a total of four Harmony HUBs in the house. up to the version 0.3.0 alpha2 all actions were presented to me as a switch in the HomeKit. Here at 0.4.0 I get only the auction of only one harmony I've shown

**Comment by materik** [comment]
@UrsLebowski so does KraigM's repo work for you? can you print your log? Thanks

**Comment by UrsLebowski** [comment]
hello, up to the version 0.3.0 the plugin was not usable for me. it crashed permanently. For all actions of all 4 HUBs switches have been created in HomeKit. but it was always only a few minutes use. Version 0.4.0 has been stable for several days now. Unfortunately, only the actions of one Harmony-HUB are shown to me.
________________________________
Von: Mattias Eriksson <notifications@github.com>
Gesendet: Montag, 5. November 2018 19:53:53
An: materik/homebridge-harmonyhub
Cc: Urs Lebowski; Mention
Betreff: Re: [materik/homebridge-harmonyhub] Multi Hub Support? (#2)


@UrsLebowski<https://github.com/UrsLebowski> so does KraigM's repo work for you? can you print your log? Thanks

**Comment by materik** [comment]
Ok. I don't feel you answered my questions. So difficult for me to debug this since I only have one hub myself.

**Comment by UrsLebowski** [comment]
> @UrsLebowski so does KraigM's repo work for you? can you print your log? Thanks

Last login: Wed Nov  7 19:02:52 on console
CC_33_3D_E3_CE_20:~ smarthome$ homebridge
[7.11.2018, 19:36:27] Loaded config.json with 0 accessories and 2 platforms.
[7.11.2018, 19:36:27] ---
[7.11.2018, 19:36:27] Loaded plugin: homebridge-harmonyhub
[7.11.2018, 19:36:27] Registering platform 'homebridge-harmonyhub.HarmonyHub'
[7.11.2018, 19:36:27] ---
[7.11.2018, 19:36:27] Loaded plugin: homebridge-itunes
[7.11.2018, 19:36:27] Registering platform 'homebridge-itunes.iTunes'
[7.11.2018, 19:36:27] ---
[7.11.2018, 19:36:28] Loaded plugin: homebridge-videodoorbell
[7.11.2018, 19:36:28] Registering platform 'homebridge-videodoorbell.Video-doorbell'
[7.11.2018, 19:36:28] ---
[7.11.2018, 19:36:28] Loading 2 platforms...
[7.11.2018, 19:36:28] [HarmonyHub] Initializing HarmonyHub platform...
[7.11.2018, 19:36:28] [iTunes] Initializing iTunes platform...
Load homebridge-videodoorbell.Video-doorbell
Setup Payload:
X-HM://0023ISYWZFEIK
Scan this code with your HomeKit app on your iOS device to pair with Homebridge:
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
                                                      
Or enter this code with your HomeKit app on your iOS device to pair with Homebridge:
                       
    ┌────────────┐     
    │ │     
    └────────────┘     
                       
[7.11.2018, 19:36:28] Homebridge is running on port xxxxx.
[7.11.2018, 19:36:28] [iTunes] Error: 34:42: execution error: „iTunes“ hat einen Fehler erhalten: „selected of AirPlay device id 234990“ kann nicht gelesen werden. (-1728)

    at ChildProcess.<anonymous> (/usr/local/lib/node_modules/homebridge-itunes/node_modules/node-osascript/lib/index.js:39:12)
    at ChildProcess.emit (events.js:182:13)
    at maybeClose (internal/child_process.js:962:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:251:5)
[7.11.2018, 19:36:28] [iTunes] Error: 34:46: execution error: „iTunes“ hat einen Fehler erhalten: „sound volume of AirPlay device id 234990“ kann nicht gelesen werden. (-1728)

    at ChildProcess.<anonymous> (/usr/local/lib/node_modules/homebridge-itunes/node_modules/node-osascript/lib/index.js:39:12)
    at ChildProcess.emit (events.js:182:13)
    at maybeClose (internal/child_process.js:962:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:251:5)
[7.11.2018, 19:36:29] [iTunes] Error: 34:42: execution error: „iTunes“ hat einen Fehler erhalten: „selected of AirPlay device id 222370“ kann nicht gelesen werden. (-1728)

    at ChildProcess.<anonymous> (/usr/local/lib/node_modules/homebridge-itunes/node_modules/node-osascript/lib/index.js:39:12)
    at ChildProcess.emit (events.js:182:13)
    at maybeClose (internal/child_process.js:962:16)
    at Socket.stream.socket.on (internal/child_process.js:381:11)
    at Socket.emit (events.js:182:13)
    at Pipe._handle.close (net.js:606:12)
[7.11.2018, 19:36:29] [HarmonyHub] {HarmonyHub} Found <1> hubs
[7.11.2018, 19:36:29] [HarmonyHub] {HarmonyHub} <2d945552-04b4-425f-8e31-e59988c5d05b> <Schlafzimmer>
[7.11.2018, 19:36:29] [iTunes] Error: 34:46: execution error: „iTunes“ hat einen Fehler erhalten: „sound volume of AirPlay device id 222370“ kann nicht gelesen werden. (-1728)

    at ChildProcess.<anonymous> (/usr/local/lib/node_modules/homebridge-itunes/node_modules/node-osascript/lib/index.js:39:12)
    at ChildProcess.emit (events.js:182:13)
    at maybeClose (internal/child_process.js:962:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:251:5)
[7.11.2018, 19:36:29] [iTunes] Error: 34:42: execution error: „iTunes“ hat einen Fehler erhalten: „selected of AirPlay device id 222369“ kann nicht gelesen werden. (-1728)

    at ChildProcess.<anonymous> (/usr/local/lib/node_modules/homebridge-itunes/node_modules/node-osascript/lib/index.js:39:12)
    at ChildProcess.emit (events.js:182:13)
    at maybeClose (internal/child_process.js:962:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:251:5)
[7.11.2018, 19:36:29] [HarmonyHub] {HarmonyHub} create new client for <Schlafzimmer>
[7.11.2018, 19:36:29] [iTunes] Error: 34:46: execution error: „iTunes“ hat einen Fehler erhalten: „sound volume of AirPlay device id 222369“ kann nicht gelesen werden. (-1728)

    at ChildProcess.<anonymous> (/usr/local/lib/node_modules/homebridge-itunes/node_modules/node-osascript/lib/index.js:39:12)
    at ChildProcess.emit (events.js:182:13)
    at maybeClose (internal/child_process.js:962:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:251:5)
[7.11.2018, 19:36:30] [HarmonyHub] {HarmonyHub} Found <3> activities: <AirPlay, Wecker stellen, AppleTV>
[7.11.2018, 19:36:31] [iTunes] Error: 34:42: execution error: „iTunes“ hat einen Fehler erhalten: „selected of AirPlay device id 222359“ kann nicht gelesen werden. (-1728)

    at ChildProcess.<anonymous> (/usr/local/lib/node_modules/homebridge-itunes/node_modules/node-osascript/lib/index.js:39:12)
    at ChildProcess.emit (events.js:182:13)
    at maybeClose (internal/child_process.js:962:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:251:5)
[7.11.2018, 19:36:31] [iTunes] Error: 34:46: execution error: „iTunes“ hat einen Fehler erhalten: „sound volume of AirPlay device id 222359“ kann nicht gelesen werden. (-1728)

    at ChildProcess.<anonymous> (/usr/local/lib/node_modules/homebridge-itunes/node_modules/node-osascript/lib/index.js:39:12)
    at ChildProcess.emit (events.js:182:13)
    at maybeClose (internal/child_process.js:962:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:251:5)
[7.11.2018, 19:36:31] [iTunes] Error: 34:42: execution error: „iTunes“ hat einen Fehler erhalten: „selected of AirPlay device id 222366“ kann nicht gelesen werden. (-1728)

    at ChildProcess.<anonymous> (/usr/local/lib/node_modules/homebridge-itunes/node_modules/node-osascript/lib/index.js:39:12)
    at ChildProcess.emit (events.js:182:13)
    at maybeClose (internal/child_process.js:962:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:251:5)
[7.11.2018, 19:36:31] [iTunes] Error: 34:46: execution error: „iTunes“ hat einen Fehler erhalten: „sound volume of AirPlay device id 222366“ kann nicht gelesen werden. (-1728)

    at ChildProcess.<anonymous> (/usr/local/lib/node_modules/homebridge-itunes/node_modules/node-osascript/lib/index.js:39:12)
    at ChildProcess.emit (events.js:182:13)
    at maybeClose (internal/child_process.js:962:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:251:5)
[7.11.2018, 19:36:31] [iTunes] Error: 34:42: execution error: „iTunes“ hat einen Fehler erhalten: „selected of AirPlay device id 222364“ kann nicht gelesen werden. (-1728)

    at ChildProcess.<anonymous> (/usr/local/lib/node_modules/homebridge-itunes/node_modules/node-osascript/lib/index.js:39:12)
    at ChildProcess.emit (events.js:182:13)
    at maybeClose (internal/child_process.js:962:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:251:5)
[7.11.2018, 19:36:32] [iTunes] Error: 34:46: execution error: „iTunes“ hat einen Fehler erhalten: „sound volume of AirPlay device id 222364“ kann nicht gelesen werden. (-1728)

    at ChildProcess.<anonymous> (/usr/local/lib/node_modules/homebridge-itunes/node_modules/node-osascript/lib/index.js:39:12)
    at ChildProcess.emit (events.js:182:13)
    at maybeClose (internal/child_process.js:962:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:251:5)

**Comment by m4d-maNu** [comment]
> 
> 
> Hello, I can report the same. I've distributed a total of four Harmony HUBs in the house. up to the version 0.3.0 alpha2 all actions were presented to me as a switch in the HomeKit. Here at 0.4.0 I get only the auction of only one harmony I've shown

Sorry for the late answer.

When I use KraigM's Harmonyhub plugin, my two hubs are found. But with this plugin comes the "MaxError"  Error.

**Comment by materik** [comment]
Ok. Cool. This log might be able to identify where the issue might be. I'll have a look as soon as possible. Might not be until the weekend however.

**Comment by twynne** [comment]
Seeing the same here.. 1 of 2 hubs detected.  Let me know if I can share any information to add to the above.

**Comment by materik** [comment]
Sorry everyone. Haven't had time to look into this. If someone else can have a look and make a PR i'm all for it!

**Comment by twynne** [comment]
> Sorry everyone. Haven't had time to look into this. If someone else can have a look and make a PR i'm all for it!

Only wish I had the skill!

One thing to add... One of my hubs (the one detected) is on firmware .193. The other is on .201

**Comment by materik** [comment]
@twynne can it be related to #8

**Comment by twynne** [comment]
> @twynne can it be related to #8

Sadly no... Only one of my hubs had updated to 206, and I downgraded it. That's the one that now works with your plugin.

**Comment by AirBorne04** [comment]
hi @materik i think i found the reason for this behavior, in [here](https://github.com/materik/homebridge-harmonyhub/blob/408d95fca33f0d29f48dfbe65437591ee112c188/lib/harmonyhub.js#L27) the promise is resolved as soon as one update event is received, therefore whichever hub gets to respond first is the one found, maybe it would be better to do a kind of timeout here to give all hubs the chance to respond.

**Comment by materik** [comment]
@AirBorne04 awesome. thanks for the advice. i had a look at this yesterday and couldn't figure it out. will have another look now!

**Comment by materik** [comment]
@m4d-maNu @twynne I've now made a fix to `develop` so if you don't mind install ```npm install -g materik/homebridge-harmonyhub#develop``` and run ```homebridge -D``` and see if that works. if you have more issues please post the log.

**Comment by VeniceNerd** [comment]
Hi guys,

I have v0.5.4 installed (fresh install) and the plugin only finds one of my two hubs. KraigM's repro used to find both. Here is a readout of my logs:

```
[2019-1-25 09:25:06] Homebridge is running on port 51464.
[2019-1-25 09:25:07] [homebridge-harmonyhub-plugin] {HarmonyHub} Found <1> hubs
[2019-1-25 09:25:07] [homebridge-harmonyhub-plugin] {HarmonyHub} <9e763bcb-4394-4e55-8c1d-fc946feff467> <Harmony Hub 2>
[2019-1-25 09:25:08] [homebridge-harmonyhub-plugin] {HarmonyHub} create new client for <Harmony Hub 2>
[2019-1-25 09:25:09] [homebridge-harmonyhub-plugin] {HarmonyHub} Found <1> activities: <Apple TV>
[2019-1-25 09:25:09] [homebridge-harmonyhub-plugin] {HarmonyHubAccessory} adding <1> activities
[2019-1-25 09:25:09] [homebridge-harmonyhub-plugin] {HarmonyHubAccessory} addActivity <Apple TV>
[2019-1-25 10:50:20] [homebridge-harmonyhub-plugin] {Service} turn <Apple TV> <on>
[2019-1-25 10:50:20] [homebridge-harmonyhub-plugin] {HarmonyHub} start activitiy: <23211844>
[2019-1-25 10:50:27] [homebridge-harmonyhub-plugin] {Service} turn <Apple TV> <off>
[2019-1-25 10:50:27] [homebridge-harmonyhub-plugin] {HarmonyHub} stop activities
[2019-1-25 10:50:38] [homebridge-harmonyhub-plugin] {Service} turn <Apple TV> <off>
[2019-1-25 10:50:38] [homebridge-harmonyhub-plugin] {HarmonyHub} stop activities
[2019-1-25 10:55:47] [homebridge-config-ui-x] Homebridge restart request received
[2019-1-25 10:55:47] [homebridge-config-ui-x] Executing restart command: killall -9 homebridge && killall -9 homebridge-config-ui-x
[2019-1-25 10:55:53] [homebridge-config-ui-x] Using Form Authentication
[2019-1-25 10:55:53] [homebridge-config-ui-x] Console v3.9.3 is listening on port 8180.
[2019-1-25 10:55:56] Loaded config.json with 0 accessories and 1 platforms.
[2019-1-25 10:55:56] ---
[2019-1-25 10:55:56] Loaded plugin: homebridge-dummy
[2019-1-25 10:55:56] Registering accessory 'homebridge-dummy.DummySwitch'
[2019-1-25 10:55:56] ---
[2019-1-25 10:55:56] Plugin /homebridge/node_modules/homebridge-harmonyhub-plugin requires Node version of >=8.12.0 which does not satisfy the current Node version of v8.10.0. You may need to upgrade your installation of Node.
[2019-1-25 10:55:57] Loaded plugin: homebridge-harmonyhub-plugin
[2019-1-25 10:55:57] Registering platform 'homebridge-harmonyhub-plugin.HarmonyHub'
[2019-1-25 10:55:57] ---
[2019-1-25 10:55:57] Loaded plugin: homebridge-config-ui-x
[2019-1-25 10:55:57] Registering platform 'homebridge-config-ui-x.config'
[2019-1-25 10:55:57] ---
[2019-1-25 10:55:57] Loading 1 platforms...
[2019-1-25 10:55:57] [homebridge-harmonyhub-plugin] Initializing HarmonyHub platform...
[2019-1-25 10:55:57] Loading 0 accessories...
Setup Payload:
X-HM://0023ISYX1CH4E
Scan this code with your HomeKit app on your iOS device to pair with Homebridge:

Or enter this code with your HomeKit app on your iOS device to pair with Homebridge:

    ┌────────────┐
    │ 031-45-157 │
    └────────────┘

[2019-1-25 10:55:57] Homebridge is running on port 51464.
[2019-1-25 10:55:58] [homebridge-harmonyhub-plugin] {HarmonyHub} Found <1> hubs
[2019-1-25 10:55:58] [homebridge-harmonyhub-plugin] {HarmonyHub} <9e763bcb-4394-4e55-8c1d-fc946feff467> <Harmony Hub 2>
[2019-1-25 10:55:59] [homebridge-harmonyhub-plugin] {HarmonyHub} create new client for <Harmony Hub 2>
[2019-1-25 10:56:00] [homebridge-harmonyhub-plugin] {HarmonyHub} Found <1> activities: <Apple TV>
[2019-1-25 10:56:00] [homebridge-harmonyhub-plugin] {HarmonyHubAccessory} adding <1> activities
[2019-1-25 10:56:00] [homebridge-harmonyhub-plugin] {HarmonyHubAccessory} addActivity <Apple TV>
```

**Comment by RienduPre** [comment]
5.3 develp version does work with multiple hubs. See my post.

The plugin writer needs to update the code to a new version and publish it

**Comment by materik** [comment]
This fix were only done on the develop branch. It's now merged to 0.5.5. Please try again and if there is still an issue I will reopen this.
